### DISCUSSION TITLE
Router Service (WORK IN PROGRESS)

### DISCUSSION THREAD

**Discussion by amcgee** [dhis2/notes#151] [Specs & RFCs]
This is a placeholder for an intra-app (and possibly inter-app) Router Service specification, including both `app-runtime` and `app-adapter` components

**Comment by amcgee** [comment]
Inspirations : https://nextjs.org/docs/basic-features/pages#pages-with-dynamic-routes

**Reply by ghost** [reply]
To me the strength of next.js' approach is the convention over configuration aspect. I've found it nice to work with. Though, I do wonder how far we'd get as we don't have a node server taking care of all the backend logic that next.js allows you to use. So for static routes I can see how we'd implement that, and it seems like a nice paradigm. Dynamic routes though I don't really see in our context just yet.

**Reply by amcgee** [reply]
@ismay can you expand on how you don't see dynamic routes in our context yet?  Dynamic routes [like `/pages/visualizations/[id].js`](https://nextjs.org/docs/routing/dynamic-routes) actually *only* works client-side in next (unless you generate the set of valid routes server-side), and I think they would actually be one of the most powerful features we'd want in a page-based routing system so that we could support wildcard id routes for an individual user, program, etc. - maybe I'm misunderstanding what you mean by dynamic routes though?

All that being said, a lot of the power (and particularly the client-side performance) of next comes from SSR and static generation - eventually we'll probably support both, but we could target (1) client-side routing only, (2) static generation / pre-hydration for known routes, and (3) SSR in that order, I think

**Reply by amcgee** [reply]
One big question for me is - should we try to rework the platform build pipeline to actually *use nextjs under the hood*, or should we just take some of the good pieces and build them ourselves?  I have concerns about both approaches, but I really do think the convention-over-configuration nextjs dev experience (@ismay agree it's one of the biggest strengths) is quite akin to what we aim to provide with our platform.

**Reply by ghost** [reply]
> can you expand on how you don't see dynamic routes in our context yet? Dynamic routes like /pages/visualizations/[id].js actually only works client-side in next (unless you generate the set of valid routes server-side), and I think they would actually be one of the most powerful features we'd want in a page-based routing system so that we could support wildcard id routes for an individual user, program, etc. - maybe I'm misunderstanding what you mean by dynamic routes though?

Yes I think we're talking about the same thing. I'm just used to the static paths/props generation with next.js, hadn't even come across the query object that the router provides. Ok, so I guess that then the only missing piece if we were to integrate something like this, is build logic that takes the filename, and converts it to parameters that are made available to the page component.

> All that being said, a lot of the power (and particularly the client-side performance) of next comes from SSR and static generation - eventually we'll probably support both, but we could target (1) client-side routing only, (2) static generation / pre-hydration for known routes, and (3) SSR in that order, I think

That would be great. Though that would mean that we need a backend that's able to handle the server-side component. I wonder how doable that is for us?

> One big question for me is - should we try to rework the platform build pipeline to actually use nextjs under the hood, or should we just take some of the good pieces and build them ourselves? I have concerns about both approaches, but I really do think the convention-over-configuration nextjs dev experience (@ismay agree it's one of the biggest strengths) is quite akin to what we aim to provide with our platform.

I would prefer to use next.js directly, if we feel their approach fits our usecase. I think it forces us to standardize, which I think is good. Plus, I don't think we have the manpower to reach anything near the level of quality that this lib provides. Yes, we might lose some customizability, but to me that's a good thing in the end. Anything custom requires manpower to do well, and I think that as an org. the functionality that next.js provides probably isn't where we want to add value.

**Reply by Mohammer5** [reply]
Should've read this before I wrote my comment.. Using next.js would be nice imo!

**Comment by amcgee** [comment]
@mediremi raised this again in #251 - copying here:

---

A common design pattern in our apps is using a grid layout with a sidebar containing navigation menu items on the left and the app content on the right. This pattern is often paired with `react-router-dom` and its `HashRouter` component to provide SPA routing.

Issues with current way of doing things
- Lack of consistency (both code and UI)
  - UI
    - Inconsistent sidebar and app content backgrounds
    - Inconsistent gutters
    - Inconsistent app content padding
    - Making menu items keyboard accessible would mean updating the version of `@dhis2/ui` for every app
  - Code
    - Different ways to organise + config vs code approach
- External app developers may not be aware of this design pattern, so may use their own components
- Forced to use `HashRouter` as backend does not know each app's routes

## Proposed solution
`d2.config.js` could be used by developers to declare the routes used by their app like so:

```js
const config = {
    // ...
    pages: {
        '/': 'src/pages/App.js',
        '/some_page/:id': 'src/pages/SomePage.js',
        // Route that shoud not visible in sidebar
        '/some_other_page': { component: 'src/pages/SomePage.js', sidebar: false }
    }
}
module.exports = config
```

and this config would then be used by the app shell to construct a sidebar and router. The runtime would reexport the `Link` and `history` exports from the routing package for consumption by the app. Declaring the routes in this way also allows the backend to build a global routing table and hence we would not need to use hash routers.

If `config.pages` is not present then the app shell would not create a sidebar and allow the app developer to manage routing.

**Reply by amcgee** [reply]
I think we need to add support for first-class platform routing soon, as this continues to come up.

I like @mediremi's suggestion, but I think instead of routes in `d2.config.js` we might be better served using a file-based route definition model similar to NextJS - basically everything under `src/pages` would automatically be turned into a route, including support for path parameters (i.e. `src/pages/visualizations/[id].js` would map to the route `/visualizations/:id`.

Additionally, I think we should probably support multiple layouts as not all apps use the sidebar pagination model - initially, the options could be `layout: 'sidebar'` or `layout: undefined` (no common app elements, just routing) with potentially others in the future like `layout: 'topbar'` or even a custom `Layout` component wrapper.

We should also probably revive something like [this](https://github.com/dhis2/ui/pull/27) to provide the building blocks for these layouts (and for pages if layout is undefined)

**Reply by amcgee** [reply]
@mediremi I'm curious what you think the backend should do with a routing table for each app... seems like it should just be configured to return the nearest `index.html` when client-side routing is configured?  I suppose if we add SSR we might want to know something about the app's routes on the backend

**Reply by mediremi** [reply]
@amcgee ~~Currently, since the backend doesn't know about each app's routing, client-side routing has to be done with hash routers (e.g. `/app/index.html#/page1`). However, if the backend knew which routes were valid we would be able to have 'proper' paths such as e.g. `/app/page1`. Returning the nearest `index.html` for any path that starts with `/app/` doesn't work if the app has static assets under its namespace.~~
**EDIT:** I was wrong - having the server serve static assets first and then fallback to `index.html` (with 404 handled by the app) as @ismay describes below is enough and the backend does not need to know about the app's routes.

**Reply by ghost** [reply]
This problem seems quite close to what you'd encounter setting up routing for SSR app. Since next.js is our inspiration here maybe we should look at how they've set up their server. It's been a while since I've set up a server for an SSR app, but I don't think there I ever needed the backend to be aware of the routes in the frontend app. I don't remember exactly what the logic I used in the server was. I believe I served static files from the public folder initially, and then fell back to serving the app (index.html) for the remaining routes (with the app being responsible for rendering a 404).

**Reply by Mohammer5** [reply]
@ismay and I are playing around with what we like in the `pages` folder, which is slighty related.
I'm testing the following right now and quite like it:

```
/pages # like next.js
/views # flat structure with all possible views
/components
/utils
```

What I like about it is:
1. views could be re-used on different pages / embedded into other views
2. `pages` remains free of "clutter", the file structure resembled the sitemap of the app, which makes it quite intuitive
3. views can still have code co-located that's relevant of that view only, preventing unnecessary abstractions
4. It's very easy to add redirects
5. Only adds a small amount of boilerplate code
6. We don't have to put view code into `components` (as it's happening in quite a few nextjs projects; at least from what I've seen so far)
7. Inside the pages folder, no one has to wonder which folder is for routes and which one is for page specific code (vercel make some good points for why they're supporting this either; [link](https://github.com/vercel/next.js/issues/3728#issuecomment-363946415))

---

Components of my current pet project look like this:

```jsx
// /pages/settings.js
import { SettingsList } from '../views/SettingsList'
export default SettingsList

// /pages/settings/[id].js
import { SettingsEdit } from '../../views/SettingsEdit'
export default SettingsEdit
```

**Reply by amcgee** [reply]
> EDIT: I was wrong - having the server serve static assets first and then fallback to index.html (with 404 handled by the app) as @ismay describes below is enough and the backend does not need to know about the app's routes.

👍 yep, this is the standard approach for SPA server config

**Reply by Mohammer5** [reply]
Btw, I'm working on the SMS Config app, I applied the structure that I've explained above: https://github.com/dhis2/sms-configuration-app/pull/253

Just to clarify: @ismay and I are just experimenting with what we'd like more and I went for the approach highlighted above. If we think that it's too much work / nesting, I'm totally open to changing it. But so far I like it quite happy with it as it allows to reduce abstractions, it's very clear where to put what and it's been extremely easy to refactor things once the structure has been set up.

**Reply by ghost** [reply]
I'm not too sure about the distinction between pages and views. I think that's a bit too much. I know I argued for pages being a place where you can have page specific code, but if we're following next.js' architecture, I think we should automate generating routes from _all_ files in pages, and just allow the rest to be in components.

 Having a separate views folder is a bit confusing to me in a react app where MVC isn't necessarily stricly enforced. To me, it could be argued that they're all views (pages and components), and that the distinction is based on their function: pages render pages, components are reusable pieces of pages. Adding views adds another category that doesn't fit the functional organisation in my opinion, and seems more related to a different concept. Ember was way closer to the MVC concept when I worked with it, with models views and controllers. For react I don't think I'd prefer using the MVC paradigm as directly when structuring the app.

**Reply by ghost** [reply]
To me it always feels like react components are more a vertical slice of the MVC pattern (as in, by concern or function), as opposed to being organized horizontally (by technology, as in: model, view or controller).

**Reply by amcgee** [reply]
+1

**Reply by Mohammer5** [reply]
Related: https://github.com/dhis2/notes/discussions/248#discussioncomment-583326

**Comment by Mohammer5** [comment]
I thought about this a bit more yesterday. I think we have the following issues if we want to provide a router service:

## We need some kind of configuration so the service know about the routes

### file structure

This makes it a bit more complicated with co-locating page-related and requires introduce a parallel structure (as I've mentioned in my previous comment). If one of our components wants to link to a route, we can't use a constant that holds the route's path and is the single source of truth for that path, which could make changing routes more complicated (although changing routes rarely happens).

### Config file

Can be read by the service easily and it's declarative, but it sounds like it's almost as much work as just specifying the router in the app. Also it'll make it slightly more difficult to change certain things based on which path is active (e. g. displaying the sidebar or not) as we're decoupling the view from the to-be-mounted components.

### Config inside source code (e. g. `src/routes.js`)

Same disadvantages as the config file

---

To me it seems this might make things more complicated than they actually help. If we just say that we routes should not be set up through (potentially dynamic) config files but instead through components as well as that all apps have to use the `react-router-dom` and its hash browser, and focus on having layout components and style consistency across apps, we'll achieve that all apps look and feel the same and getting into the routes is not really a biggie as you just have to read through the router component that the app implements .

Wrapping `react-router-dom` or creating our own service seems like it's either an unnecessary abstraction (as we either have to expose the whole api surface of `react-router-dom` or restrict the custom service's api) or we'll end up implementing a router service which is just re-inventing the wheel.

UI inconsistencies is something we might want as apps might have different requirements / a different target audience, so having the freedom of choosing is nice. The only thing I think we should disallow is having a config for the routes but use the same approach in all apps (component based routing instead of config based routing that dynamically mounts the route components)

**Reply by Mohammer5** [reply]
If we want to offer SSR, then this is a whole different issue though..

**Comment by HendrikThePendric** [comment]
> A common design pattern in our apps is using a grid layout with a sidebar containing navigation menu items on the left and the app content on the right. This pattern is often paired with react-router-dom and its HashRouter component to provide SPA routing.

I agree this is a very common pattern and some form of standardisation / automation would be good. Personally I would probably just prefer to implement some components that facilitate this common layout pattern instead of introducing a more rigid configuration based service.
